test/integration/swaps.ts

```diff
@@ -0,0 +1,148 @@
+
+    const path = [weth.address, bdeu.address, bdx.address];
+
+    weth.approve(router.address, amountIn);
+    const currentBlock = await hre.ethers.provider.getBlock("latest");
+    await router.swapExactTokensForTokens(amountIn, 0, path, deployer.address, currentBlock.timestamp + 1e5);
+  });
+
+  it("Should choose best path (lenght 2)", async () => {
+    const deployer = await getDeployer(hre);
+
+    const bdx = await getBdx(hre);
+    const weth = await getWeth(hre);
+    const router = await getUniswapRouter(hre);
+
+    const amountIn = to_d18(0.0001);
+
+    const pathsPrices = await generatePaths(amountIn, weth.address, bdx.address);
+    console.log(
+      pathsPrices.map(x => ({
+        path: x.path,
+        amountOut: d18_ToNumber(x.amountOut)
+      }))
+    );
+
+    const bestPath = await chooseBestPath(pathsPrices);
+    console.log(d18_ToNumber(bestPath.amountOut));
+    expect(bestPath.path.length).to.eq(2);
+
+    weth.approve(router.address, amountIn);
+    const currentBlock = await hre.ethers.provider.getBlock("latest");
+    await router.swapExactTokensForTokens(amountIn, 0, bestPath.path, deployer.address, currentBlock.timestamp + 1e5);
+  });
+
+  it("Should choose best path (length 3)", async () => {
+    const deployer = await getDeployer(hre);
+
+    const bdx = await getBdx(hre);
+    const weth = await getWeth(hre);
+    const router = await getUniswapRouter(hre);
+
+    const currentBlock = await hre.ethers.provider.getBlock("latest");
+
+    // sabotage the price on the direct swap, so the algorithm prefers an indirect path
+    const amountInForSabotage = to_d18(10);
+    console.log("amountInForSabotage:", d18_ToNumber(amountInForSabotage));
+    await weth.approve(router.address, amountInForSabotage);
+    await router.swapExactTokensForTokens(amountInForSabotage, 0, [weth.address, bdx.address], deployer.address, currentBlock.timestamp + 1e5);
+
+    const amountIn = to_d18(0.0001);
+
+    const pathsPrices = await generatePaths(amountIn, weth.address, bdx.address);
+    console.log(
+      pathsPrices.map(x => ({
+        path: x.path,
+        amountOut: d18_ToNumber(x.amountOut)
+      }))
+    );
+
+    const bestPath = await chooseBestPath(pathsPrices);
+    console.log(d18_ToNumber(bestPath.amountOut));
+    expect(bestPath.path.length).to.eq(3);
+
+    weth.approve(router.address, amountIn);
+    await router.swapExactTokensForTokens(amountIn, 0, bestPath.path, deployer.address, currentBlock.timestamp + 1e5);
+  });
+
+  async function getTokensAvailebleForSwap() {
+    return [
+      ...(await getAllBDStables(hre)).map(x => x.address),
+      (await getBdx(hre)).address,
+      (await getWbtc(hre)).address,
+      (await getWeth(hre)).address
+    ];
+  }
+
+  async function generatePaths(amountIn: BigNumber, addressIn: string, addressOut: string): Promise<{ path: string[]; amountOut: BigNumber }[]> {
+    const availableTokens = await getTokensAvailebleForSwap();
```

Comment:
we should have those addresses already so `getTokensAvailebleForSwap()` will be redundant on FE

Reply:
yep, this why I encapsulated it in a separate function :)